### PR TITLE
[serialize] Support Avgpool2D with count_include_pad=False

### PR DESCRIPTION
**Pull request by dayo09**
Let's support AvgPool2D with count_include_pad=False partially.

TICO-DCO-1.0-Signed-off-by: Dayoung Lee <dayoung.lee@samsung.com>

### CODE REVIEW

File: tico/serialize/operators/op_avg_pool2d.py [hunk 1]
```diff
@@ -26,16 +27,79 @@
...
 
 @register_node_visitor
 class AvgPool2DVisitor(NodeVisitor):
+    """
+    This class defines how to serialize AvgPool2D operation into Circle IR.
+
+    Torch                           | Circle
+
+    count_include_pad: True/False   | (count_include_pad): Always False
+    (padding): Always "valid"       | padding: "valid"/"same"
```
Comment (jinevening) [inline tico/serialize/operators/op_avg_pool2d.py:42]: Torch avgpool2d Op can have padding. It's not always "valid".

https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.avg_pool2d.html

File: tico/serialize/operators/op_avg_pool2d.py [hunk 2]
```diff
@@ -26,16 +27,79 @@
...
+            return False
+        else:
+            return True
+
+    def has_same_padding(self, args: AvgPool2dArgs) -> bool:
+        input_shape = list(extract_shape(args.input))
+        kernel_size = args.kernel_size
+        stride = args.stride
+        padding = args.padding
+
```
Comment (jinevening) [inline tico/serialize/operators/op_avg_pool2d.py:65]: The below equation only works when ceil_mode == False. Better to add a guard here

```suggestion
        # TODO Update this function when supporting ceil_mode = True
        assert args.ceil_mode is False
```

File: tico/serialize/operators/op_avg_pool2d.py [hunk 3]
```diff
@@ -26,16 +27,79 @@
...
+            (input_shape[1] + padding[0] * 2 - kernel_size[0]) / stride[0] + 1
+        )
+        output_width = math.floor(
+            (input_shape[2] + padding[1] * 2 - kernel_size[1]) / stride[1] + 1
+        )
+
+        if input_shape[1] == output_height and input_shape[2] == output_width:
+            return True
+        else:
+            return False
```
Comment (jinevening) [inline tico/serialize/operators/op_avg_pool2d.py:76]: (optional) For simplicity

```suggestion
        return input_shape[1] == output_height and input_shape[2] == output_width
```

File: test/modules/op/avg_pool2d.py [hunk 4]
```diff
@@ -96,3 +98,49 @@ def forward(self, tensor):
 
     def get_example_inputs(self):
         return (torch.randn(2, 4, 8, 16),)
+
+
+class AvgPoolWithNoCountIncludePad(torch.nn.Module):
```
Comment (jinevening) [inline test/modules/op/avg_pool2d.py:103]:
```suggestion
class AvgPoolWithSamePaddingNoCountIncludePad(torch.nn.Module):
```

File: tico/serialize/operators/op_avg_pool2d.py [hunk 5]
```diff
@@ -81,32 +144,30 @@ def define_padding_node():
...
+            # Don't care count_include_pad
+            result = self.define_avgpool_node(
+                [avgpool_input], [node], "VALID", stride, kernel_size
+            )
+        elif count_include_pad is True:
+            # Add padding before avgpool2d
+            # Circle's avgpool2d does not support count_include_pad=True, so we need to add padding manually
             avgpool_input = define_padding_node()
+            result = self.define_avgpool_node(
+                [avgpool_input], [node], "VALID", stride, kernel_size
```
Comment (Seunghui98) [inline tico/serialize/operators/op_avg_pool2d.py:157]: May I ask a question? 🙂
if count_include_pad is True, Why we set "VALID"...?

Comment (dayo09) [reply tico/serialize/operators/op_avg_pool2d.py:157]: Let me share this psuedo code for better understanding!


```py
if count_include_pad == True:
    (Circle cannot represent count_include_pad=True in AvgPool2D. Therefore we manually add zero padding node.)
    DEFINE zero padding node
    DEFINE avgpool node with no padding (valid)
if count_include_pad == False:
    (Lucky! Circle can represent count_include_pad=False)
    define avgpool node with same/valid padding.
    
    (However, it cannot represent all paddings. So, if the padding is not same or valid, we throw an error.)
    if the paddding is neither same nor valid, throw an error.
```

Comment (dayo09) [reply tico/serialize/operators/op_avg_pool2d.py:168]: Plus, let me revise the statements.

**Review by jinevening — APPROVED** [review]
LGTM

File: tico/serialize/operators/op_avg_pool2d.py (hunk 5 above)
Comment (Seunghui98) [reply tico/serialize/operators/op_avg_pool2d.py:168]: @dayo09 
Ah! I understood it 😀
Thanks for letting me know 👍
